File: samples/client/3_0_3_unit_test/java/src/main/java/org/openapijsonschematools/client/schemas/validation/ItemsValidator.java
package org.openapijsonschematools.client.schemas.validation;

import org.checkerframework.checker.nullness.qual.Nullable;

import java.util.ArrayList;
import java.util.List;

public class ItemsValidator implements KeywordValidator {
    public final Class<? extends JsonSchema> items;

    public ItemsValidator(Class<? extends JsonSchema> items) {
        this.items = items;
    }

    @Override
    public @Nullable PathToSchemasMap validate(
        JsonSchema schema,
        @Nullable Object arg,
        ValidationMetadata validationMetadata,
        @Nullable List<PathToSchemasMap> containsPathToSchemas,
        @Nullable PathToSchemasMap patternPropertiesPathToSchemas
    ) {
        if (!(arg instanceof List)) {
            return null;
        }
        PathToSchemasMap pathToSchemas = new PathToSchemasMap();
        // todo add handling for prefixItems
        int i = 0;
        for(Object itemValue: (List<?>) arg) {
            List<Object> itemPathToItem = new ArrayList<>(validationMetadata.pathToItem());
            itemPathToItem.add(i);
            ValidationMetadata itemValidationMetadata = new ValidationMetadata(
                    itemPathToItem,
                    validationMetadata.configuration(),
                    validationMetadata.validatedPathToSchemas(),
                    validationMetadata.seenClasses()
            );
            JsonSchema itemsSchema = JsonSchemaFactory.getInstance(items);
            if (itemValidationMetadata.validationRanEarlier(itemsSchema)) {
                // todo add_deeper_validated_schemas
                i +=1;
                continue;
            }
            PathToSchemasMap otherPathToSchemas = JsonSchema.validate(itemsSchema, itemValue, itemValidationMetadata);
            pathToSchemas.update(otherPathToSchemas);
            i += 1;
        }
        return pathToSchemas;
    }
}

File: samples/client/3_1_0_unit_test/kotlin/src/main/kotlin/org/openapijsonschematools/client/schemas/validation/ItemsValidator.kt
package org.openapijsonschematools.client.schemas.validation

import org.openapijsonschematools.client.exceptions.ValidationException

class ItemsValidator : KeywordValidator {
    @Throws(ValidationException::class)
    override fun validate(
        data: ValidationData
    ): PathToSchemasMap? {
        val items: Class<out JsonSchema<*>> = data.schema.items ?: return null
        if (data.arg !is List<*>) {
            return null
        }
        if (data.arg.isEmpty()) {
            return null
        }
        val pathToSchemas = PathToSchemasMap()
        val minIndex = if (data.schema.prefixItems != null) data.schema.prefixItems.size else 0
        val itemsSchema = JsonSchemaFactory.getInstance(items)
        for (i in minIndex until data.arg.size) {
            val itemPathToItem: MutableList<Any> = ArrayList(data.validationMetadata.pathToItem)
            itemPathToItem.add(i)
            val itemValidationMetadata = ValidationMetadata(
                itemPathToItem,
                data.validationMetadata.configuration,
                data.validationMetadata.validatedPathToSchemas,
                data.validationMetadata.seenClasses
            )
            if (itemValidationMetadata.validationRanEarlier(itemsSchema)) {
                // todo add_deeper_validated_schemas
                continue
            }
            val otherPathToSchemas = JsonSchema.validate(itemsSchema, data.arg[i], itemValidationMetadata)
            pathToSchemas.update(otherPathToSchemas)
        }
        return pathToSchemas
    }
}
